TELCODOCS#1831: Local storage overview (#73772)
openshift/openshift-docs: modules/comparison-of-solutions-to-provision-node-local-storage.adoc

|Support for block storage |Yes |Yes |No

|Support for file storage |Yes |Yes |Yes

|Support for object storage ^[1]^ |No |No |No

|====
[.small]
--
1. None of the solutions ({lvms}, LSO, and HPP) provide support for object storage. Therefore, an S3 object storage solution, such as `MultiClusterGateway` from the Red Hat OpenShift Data Foundation, is needed. All of the solutions can serve as underlying storage providers for the S3 object storage solutions.
--

[id="comparing-core-functionalities"]
== Comparison of the support for core functionalities

The following table compares how {lvms}, LSO, and HPP support core functionalities for provisioning local storage:

.Comparison of the core functionalities of {lvms}, LSO, and HPP to provision local storage
[cols="6a,5a,5a,5a",options="header"]
|====

|Functionality|LVM Storage|Local Storage Operator (LSO)|HostPath Provisioner (HPP)

|Support for automatic file system formatting |Yes |Yes |N/A

|Support for dynamic provisioning |Yes |No |No

|Support for using software Redundant Array of Independent Disks (RAID) arrays 
|Yes 

Supported on 4.15 and higher versions.

|Yes |Yes

|Support for transparent disk encryption |Yes 

> Supported with 4.16 version.

Reviewer: ```suggestion
Supported with 4.16 and higher versions.
```

Author: Updated as per the suggestion.

|Support for volume snapshots and volume clones |Yes |No |No

|Support for thin provisioning |Yes

Devices are thin-provisioned by default.

|Yes

You can configure the devices to point to thin-provisioned volumes

|Yes

You can configure a path to point to thin-provisioned volumes.

|Support for automatic disk discovery and setup |Yes

Automatic disk discovery is available during installation and runtime. You can also dynamically add the disks to the storage classes to increase the storage capacity.

|Technology Preview

Automatic disk discovery is available during installation. 

|No

|====

[id="comparing-performance-and-isolation-boundary"]
== Comparison of performance and isolation capabilities

The following table compares the performance and isolation capabilities of {lvms}, LSO, and HPP in provisioning local storage.

.Comparison of the performance and isolation capabilities of {lvms}, LSO, and HPP to provision local storage
[cols="5a,6a,6a,6a",options="header"]
|====

|Functionality|LVM Storage|Local Storage Operator (LSO)|HostPath Provisioner (HPP)

|Performance

|I/O speed is shared for all workloads that use the same storage class. 

> Furthermore, block storage allows direct I/O operations. However, thin-provisioning can affect the performance. 

Reviewer: ```suggestion
Block storage allows direct I/O operations. Thin-provisioning can affect performance. 
```

Author: Updated as per the suggestion.

|I/O speed is shared for all workloads that use the same storage class. 

Furthermore, block storage allows direct I/O operations.

> |I/O speed is shared for all workloads that use the same storage class. However, the restrictions imposed by the underlying filesystem can affect the performance.

Reviewer: ```suggestion
|I/O speed is shared for all workloads that use the same storage class. The restrictions imposed by the underlying filesystem can affect performance.
```

Author: Updated as per the suggestion.

|Isolation boundary ^[1]^

|LVM Logical Volume (LV)

It provides higher level of isolation compared to HPP.

|LVM Logical Volume (LV)

It provides higher level of isolation compared to HPP

|Filesystem path

It provides lower level of isolation compared to LSO and {lvms}.

|====
[.small]
--
1. The isolation boundary refers to the level of separation between different workloads or applications that use local storage resources.
--

[id="comparing-additional-functionalities"]
== Comparison of the support for additional functionalities

The following table compares the additional features provided by {lvms}, LSO, and HPP to provision local storage:

.Comparison of the additional functionalities of {lvms}, LSO, and HPP to provision local storage
[cols="6a,5a,5a,5a",options="header"]
|====

|Functionality|LVM Storage|Local Storage Operator (LSO)|HostPath Provisioner (HPP)

> |Available filesystems |`ext4`, `xfs` |`ext4`, `xfs` |Any mounted system available on the node is supported.

Reviewer: My two cents - I think this should be moved up to its own section and/or included with the storage type table. I wouldn't consider this an additional functionality as this would likely be a very common question that would need to be answered.

Author: Moved it to "Comparison of the support for storage types and filesystems" section.

> |Support for CSI driver |Yes |No |No

Reviewer: It's not entirely clear to me what we're describing with this one so more discussion may be needed in terms of how a user/customer would interpret this. For example, is this "support for replacing the CSI driver", "product uses a CSI driver", etc?

Author: @jakobmoellerdev @DanielFroehlich 
Could you PTAL?

Reviewer 2: So For me LVMS IS the CSI Driver. That means that we can either add an option similar to "Is a CSI Driver" or leave it out completely. No strong opinion there. Just support for CSI driver is misleading I believe.

Author: @jakobmoellerdev 
We have the following line in the intro to LVM Storage:

"Logical Volume Manager Storage uses the TopoLVM CSI driver.."

Based on this, how about modifying "Support for CSI driver" to "Support for integrating with the CSI driver"?

Reviewer 2: I think this is ultimately irrelevant to the user. A user will only be interested if the storage class in kubernetes is provisioned by a CSI Driver and wether it supports dynamic provisioning. As such I think we could just remove this one here.

Author: Removed the "Support for CSI driver" row.

|Support for generic ephemeral volumes |Yes |No |No

|Support for CSI inline ephemeral volumes |No |No |No

|Support for storage topology |Yes 

Supports CSI node topology 

|Yes

> However, LSO provides partial support for storage topology through node tolerations. 

Reviewer: ```suggestion
LSO provides partial support for storage topology through node tolerations. 
```

Author: Updated as per the suggestion.

// Module included in the following assemblies:
//

> // * storage/persistent_storage/persistent_storage_local/overview-and-comparison-of-local-storage-options

Reviewer 3: Since you're already going to make changes in this file, I'd update this entry to have the correct file name:
```suggestion
// * storage/persistent_storage/persistent_storage_local/ways-to-provision-local-storage.adoc
```

Author: Updated as per the suggestion.

:_mod-docs-content-type: CONCEPT
[id="comparison-of-solutions-to-provision-node-local-storage_{context}"]

> = Comparison of LVM Storage, LSO, and HPP

Reviewer 3: This is not the biggest deal but I think it looks a little awkward to have this heading followed by another heading with no text in between. If you can, could you add some content here? Here's one simple sentence:
```suggestion
= Comparison of LVM Storage, LSO, and HPP

The following sections compare different aspects of the supported storage types.
```

Author: Added the following info:

"The following sections compare the functionalities provided by {lvms}, Local Storage Operator (LSO), and HostPath Provisioner (HPP) to provision local storage."

> [id="comparing-storage-types"]

Reviewer 3: Per our [docs guidelines](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#anchoring-in-module-files), each anchor ID in a module needs a context at the end of it:
```suggestion
[id="comparing-storage-types_{context}"]
```

Author: Added context to the end of all Ids.

== Comparison of the support for storage types and filesystems

The following table compares the support for storage types and filesystems provided by {lvms}, Local Storage Operator (LSO), and HostPath Provisioner (HPP) to provision local storage:

.Comparison of the support for storage types and filesystems
[cols="6a,5a,5a,5a",options="header"]
|====

|Functionality|LVM Storage|LSO |HPP

|Support for block storage |Yes |Yes |No

|Support for file storage |Yes |Yes |Yes

|Support for object storage ^[1]^ |No |No |No

|Available filesystems |`ext4`, `xfs` |`ext4`, `xfs` |Any mounted system available on the node is supported.

|====
[.small]
--

> 1. None of the solutions ({lvms}, LSO, and HPP) provide support for object storage. Therefore, an S3 object storage solution, such as `MultiClusterGateway` from the Red Hat OpenShift Data Foundation, is needed. All of the solutions can serve as underlying storage providers for the S3 object storage solutions.

Reviewer 4: ```suggestion
1. None of the solutions ({lvms}, LSO, and HPP) provide support for object storage. Therefore, if you want to use object storage, you need an S3 object storage solution, such as `MultiClusterGateway` from the Red Hat OpenShift Data Foundation. All of the solutions can serve as underlying storage providers for the S3 object storage solutions.
```

Author: Updated as per the suggestion.

> 1. None of the solutions ({lvms}, LSO, and HPP) provide support for object storage. Therefore, if you want to use object storage, you need an S3 object storage solution, such as `MultiClusterGateway` from the Red Hat OpenShift Data Foundation. All of the solutions can serve as underlying storage providers for the S3 object storage solutions.

Reviewer 3: Same comment about NBSP for Red Hat:
```suggestion
1. None of the solutions ({lvms}, LSO, and HPP) provide support for object storage. Therefore, if you want to use object storage, you need an S3 object storage solution, such as `MultiClusterGateway` from the Red{nbsp}Hat OpenShift Data Foundation. All of the solutions can serve as underlying storage providers for the S3 object storage solutions.
```

Author: Updated as per the suggestion.

--

> [id="comparing-core-functionalities"]

Reviewer 3: same comment about module IDs:
```suggestion
[id="comparing-core-functionalities_{context}"]
```

Author: Updated as per the suggestion.

== Comparison of the support for core functionalities

The following table compares how {lvms}, Local Storage Operator (LSO), and HostPath Provisioner (HPP) support core functionalities for provisioning local storage:

.Comparison of the support for core functionalities
[cols="6a,5a,5a,5a",options="header"]
|====

|Functionality|LVM Storage|LSO |HPP

|Support for automatic file system formatting |Yes |Yes |N/A

|Support for dynamic provisioning |Yes |No |No

|Support for using software Redundant Array of Independent Disks (RAID) arrays 
|Yes 

> Supported on 4.15 and higher versions.

Reviewer 3: [Per ISG](https://www.ibm.com/docs/en/ibm-style?topic=references-product-versions#references-to-earlier-or-later-versions), the way to reference later versions is by saying "or later":
```suggestion
Supported on 4.15 or later.
```

Author: Updated as per the suggestion.

|Yes |Yes

|Support for transparent disk encryption |Yes 

> Supported with 4.16 and higher versions.

Reviewer 3: Same comment as above, also suggestion to change to say "supported on" for consistency with the other entry but that part is up to you:
```suggestion
Supported on 4.16 or later.
```

Author: Updated as per the suggestion.

|Yes |Yes

|Support for volume based disk encryption |No |No |No

|Support for disconnected installation |Yes |Yes |Yes

|Support for PVC expansion |Yes |No |No

|Support for volume snapshots and volume clones |Yes |No |No

|Support for thin provisioning |Yes

Devices are thin-provisioned by default.

|Yes

You can configure the devices to point to thin-provisioned volumes

|Yes

You can configure a path to point to thin-provisioned volumes.

|Support for automatic disk discovery and setup |Yes

> Automatic disk discovery is available during installation and runtime. You can also dynamically add the disks to the storage classes to increase the storage capacity.

Reviewer 5: I'm not the LVMS expect, could anyone from LVMS team to confirm:

You can also dynamically add the disks to the `storage classes` to increase the storage capacity.
or should be:
You can also dynamically add the disks to the `LVMCluster` to increase the storage capacity.

Author: Hi @duanwei33 
Yes, that's right! However, upon configuring the device classes (where we add disks) in the LVMCluster CR, LVM Storage automatically creates a storage class for each device class. These storage classes can be specified in the PVCs. 

I think saying "add the disks to LVMCluster CR" will be very specific. But using "storage classes" provides more context I guess. 

@jakobmoellerdev 
WDYT?

Reviewer 2: I think @duanwei33 is correct, but you have a point. we can respecify to 

> You can also dynamically add the disks to LVMCluster to increase the storage capacity of existing storage classes.

Author: This sounds great!
Updated as per the suggestion.

|Technology Preview

Automatic disk discovery is available during installation. 

|No

|====

> [id="comparing-performance-and-isolation-boundary"]

Reviewer 3: Same comment about module IDs:
```suggestion
[id="comparing-performance-and-isolation-boundary_{context}"]
```

== Comparison of performance and isolation capabilities

The following table compares the performance and isolation capabilities of {lvms}, Local Storage Operator (LSO), and HostPath Provisioner (HPP) in provisioning local storage.

.Comparison of performance and isolation capabilities
[cols="5a,6a,6a,6a",options="header"]
|====

|Functionality|LVM Storage|LSO |HPP

|Performance

|I/O speed is shared for all workloads that use the same storage class. 

Block storage allows direct I/O operations. 

Thin-provisioning can affect the performance. 

> |I/O speed is shared for all workloads that use the same storage class. 

Reviewer 5: (I might be wrong, but) If the provisioning is based on disks(not partition), I think the I/O is not shared？

Author: @jakobmoellerdev @brandisher 
Could you PTAL?

Reviewer 2: For thin provisioning based on volume group, I/O is shared either way. This is because one storage class maps to one volume group. This volume group will get I/O from all pods that a have a volume mounted on it. 

However, this does not directly translate to shared I/O from the pods: a linear volume group might have one pod use data blocks from disk 1 and one pod from disk 2 so they might not share I/O in that layout. 

TLDR: under certain circumstances I/O is not shared, but we do not want to expose this abstraction to the users so it is safer to say that I/O is shared for all workloads (even if it really isnt under some border cases)

Author: @duanwei33 
For LSO, how about changing it to "I/O speed is not shared for workloads."
Should we add more info to provide better context?

Hi @gcharot ,
Could you PTAL?
Thanks!

Reviewer 2: Just to understand correctly, I was expecting LSO to use LocalVolume to setup a StorageClass. Within this storage class you use one to many devices. However in this storageclass, I/O between workloads is shared. If there is only one disk then to avoid sharing for workloads, there must be one sc per workload, right?

Reviewer 5: Hi @jakobmoellerdev, local storage could only provide static provisioning, we could define disks or partitions in LSO CR, and it transfers each disk or partition to one pv (the pv number based on how many disks/partitions are detected), and here the storageclass is also created but it looks like a label to get bound with a pv when using the PVC inside the workload.  
So one common use case is:
- user creates PVC1 with storageclass (created by LSO CR), the PVC1 get bound with PV1(disk1) and it is consumed by workload1
- user creates PVC2 with the same storageclass,  the PVC2 get bound with PV2(disk2) and it is consumed by workload2

Here workload1 and workload2 are using two different disks, so I guess the I/O is not shared.

> I/O speed is shared for all workloads that use the same storage class. 

cc @jsafrane @gnufied May I get some help/clarification if I wrote something wrong :(

Author: Updated it to "I/O depends on LSO configuration" as suggested by the QE @duanwei33

Block storage allows direct I/O operations.

|I/O speed is shared for all workloads that use the same storage class. 

> The restrictions imposed by the underlying filesystem can affect the performance.

Reviewer 4: >  can affect the performance
Will the typical user know _what_ performance can be affected?

Author: I think in this context we are referring to "I/O speed". So it means ".. underlying filesystem can affect the I/O speed."

But I'll let @jakobmoellerdev comment on this!

Reviewer 5: It looks good to me :)

|Isolation boundary ^[1]^

|LVM Logical Volume (LV)

It provides higher level of isolation compared to HPP.

|LVM Logical Volume (LV)

It provides higher level of isolation compared to HPP

|Filesystem path

It provides lower level of isolation compared to LSO and {lvms}.

|====
[.small]
--
1. Isolation boundary refers to the level of separation between different workloads or applications that use local storage resources.
--

> [id="comparing-additional-functionalities"]

Reviewer 3: Same comment about module IDs:
```suggestion
[id="comparing-additional-functionalities_{context}"]
```

Author: Updated as per the suggestion.
